third_party/move/move-vm/runtime/src/loader/function.rs

```diff
@@ -237,29 +245,19 @@ impl LazyLoadedFunction {
         })
     }
 
-    /// Resolves a function into a loaded function. This verifies existence of the named
-    /// function as well as whether it has the type used for deserializing the captured values.
-    fn resolve<RTTCheck: RuntimeTypeCheck>(
-        module_storage: &impl ModuleStorage,
+    /// Verifies that the function argument types match the layouts used for deserialization. This
+    /// is only done in paranoid mode. Because of integrity of storage and the guarantees of public
+    /// functions, these checks should not fail.
+    /// Note that because we need to charge gas for modules loaded during layout construction, the
+    /// only "paranoid" check is the comparison of layouts.
```

Comment:
as opposed to what? loading modules? or layouts?

Reply:
I mean that we do a paranoid check, but check loads modules => needs to be metered. Which means paranoid mode is no longer free which is not what we want (paranoid mode can be on/off, but should not incur gas costs by itself). Hence here to load layouts we need to charge gas before the paranoid check.

In any case, I think it is easier to get rid of this check as it does not really save us from anything it seems, as @wrwg pointed out. PR: #16791